### DISCUSSION TITLE
What KP matches does the databaseviewer display?

### DISCUSSION THREAD

**Discussion by mattiasmar** [introlab/rtabmap#1074] [Q&A] [answered]
Does the databaseviewer display keypoint matches found during creation of the db file or does it recalculate the matches on its own?

**Comment by matlabbe** [comment, marked as answer]
They are the keypoints saved in the database. It can re-extract them on specific post processing actions (when trying different parameters tham when the db was created).

**Reply by mattiasmar** [reply]
Is there a command to update the keypoint extraction and matching of one particular image pair?

**Reply by matlabbe** [reply]
No, individual post-processing actions cannot be saved back in database. THe workflow I use is to try different parameters with `rtabmap-databaseViewer` on some frames, then when satisfied, reprocess the whole database with those parameters (example here by increasing the number of features used for loop closure detection):
```
rtabmap-reprocess --Kp/MaxFeatures 800 rtabmap.db output.db
```

**Reply by mattiasmar** [reply]
My previous question was about _how_ one can "try different parameters with rtabmap-databaseViewer on some frames".
I see a window ("info") that shows all the parameters that were used:
<img width="463" alt="image" src="https://github.com/introlab/rtabmap/assets/48177580/d6cdb2b4-9775-48be-9357-1c0f4e511cf4">
I don't see where I can edit one or more parameters and then re-execute keypoint detection (and matching). Where do I do that?

**Reply by mattiasmar** [reply]
...you probably meant the "Add" button under in the "Constraint view" pane.
<img width="363" alt="image" src="https://github.com/introlab/rtabmap/assets/48177580/ad61cb19-2b6d-4065-be90-efb7285fe2e9">

**Reply by mattiasmar** [reply]
But where do one change parameters before clicking on Add? For example keypoint threshold, match threshold, MaxFeatures, where can I _edit_ those in the databaseviewer?

**Reply by matlabbe** [reply]
Open the "Core Parameters" view in databaseviewer. To re-extract features, check parameter `RGBD/LoopClosureReextractFeatures`
![image](https://github.com/introlab/rtabmap/assets/2319645/9f536019-e1a8-4dfe-853e-88d08c6954a8)

Yes, use Add button to add a new loop closure between two selected frames.
